datalad/support/gitrepo.py

```diff
@@ -2201,8 +2208,10 @@ def _call_git(self, args, files=None, expect_stderr=False, expect_fail=False,
         expect_fail : bool, optional
           A non-zero exit is expected and should not be elevated above the
           DEBUG level.
-        check_fake_dates : bool, optional
-          If enabled (default), inspect if dates should be faked for dataset
+        read_only : bool, optional
+          If False (default), do not assume that command does not introduce changes and
```

Comment:
I find the double negative hard to parse here.  When trying to come up with a rewrite, I stumbled on how to phrase it accurately.  I started to think about the larger issue with `read_only`: some commands in git are semantically read-only but actually modify the repo in ways that can clash with other running git commands.  For example, `status` and `diff` will refresh and write the index.  These writes are optimizations to avoid recomputing the same thing on the next git call.

In the case of `status`, the global `--no-optional-locks` option can be used to avoid these.  27344d6a6c (git: add --no-optional-locks option, 2017-09-27) makes it clear that this option doesn't cover (at least yet) all the commands that do such an "opportunistic" refresh.

So, while I think `read_only` (versus a more specific `check_fake_dates`) is appealing because it adds a single high-level knob that conceptually covers `check_fake_dates` (and now locking), my concern is that it's misleading because it doesn't reflect how git behaves underneath.

Anyway, perhaps not a reason to avoid adding `read_only`, but I think it's worth thinking about and making the docstring clearer.  Also, given the support for parallel operations now, our `status` calls in `.dirty` should use `--no-optional-locks`, and we should audit our `diff` calls to see if we can switch to a lower-level command (`diff-files`, ...) that doesn't refresh the index.  (But neither of those things need to be tied to this PR of course.)

As an unrelated note, this parameter should be documented in the other `call_git*` commands, perhaps with some fancy formatting like config's `_where_reload`, though I think nearby comments reminding future updaters to cover the other spots would do.

Reply:
yeah, I was not 100% happy with read_only as well, but other ones I thought about are either not generic enough (e.g, `no_commit`) or vague (e.g. `changes_state`).
Please suggest changes to the wording - I did think about making it also more specific, like "no commits etc", but just makes it longer for no good value.  We could add a disclaimer/note that "Note: even if 'read-only' commands could refresh index etc." may be, but again -- not 100% certain if such level of detail is really needed here

Reply:
adjusted description a bit, hopefully it is better

Reply:
> I did think about making it also more specific, like "no commits etc", but just makes it longer for no good value.

Or it actually makes it incorrect.  The problem isn't that the parameter name is inaccurate; it's that the caller may think some commands are clearly read-only, but in fact they write to the repository.  In those cases, it wouldn't be appropriate to skip the locks needed for parallel operations.